### PR TITLE
Add metrics for elastic task assignment

### PR DESCRIPTION
**Pull request by somandal**
This PR adds metrics for elastic task assignment. The metrics track a) actual number of partitions/task, b) whether the actual partitions/task is above the configured partitionsPerTask.

It does this by having the StickyPartitionAssignmentStrategy implement the MetricsAware interface and pulling these into the Coordinator metrics.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 1]
```diff
@@ -470,6 +499,24 @@ protected int constructExpectedNumberOfTasks(DatastreamGroup dg, List<String> in
     return expectedNumberOfTasks;
   }
 
+  private void updateOrRegisterElasticTaskAssignmentMetrics(DatastreamGroupPartitionsMetadata datastreamPartitions,
+      int totalTaskCount) {
+    int totalPartitions = datastreamPartitions.getPartitions().size();
+    int actualPartitionsPerTask = (totalTaskCount / totalPartitions)
```
Comment (jzakaryan) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java]: `totalTaskCount` and `totalPartitions` need to be flipped, otherwise `actualPartitionsPerTask` will always resolve to 0 or 1.

File: datastream-testcommon/src/main/java/com/linkedin/datastream/testutil/DatastreamTestUtils.java [hunk 2]
```diff
@@ -197,7 +197,6 @@ public static void updateDatastreams(ZkClient zkClient, String cluster, Datastre
    */
   public static void deleteDatastreamsNumTasks(ZkClient zkClient, String cluster, String... datastreams) {
     for (String datastreamName : datastreams) {
-      zkClient.ensurePath(KeyBuilder.datastreams(cluster));
```
Comment (jzakaryan) [inline datastream-testcommon/src/main/java/com/linkedin/datastream/testutil/DatastreamTestUtils.java:200]: I'm assuming this change is unrelated to the current PR. Just cleanup?

**Review by jzakaryan — DISMISSED** [review]
LGTM

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 3]
```diff
@@ -592,4 +639,39 @@ private int getNumTasksForDatastreamFromZK(String stream) {
 
     return Integer.parseInt(_zkClient.readData(numTasksPath));
   }
+
+  /**
+   * Register metrics related to elastic task assignment
+   */
+  private void registerElasticTaskAssignmentMetrics(String taskPrefix) {
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:646]: Since the gauge metrics is registered per datastream, don't we need a deregister path when there are no tasks associated with the datastream prefix on this host.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 4]
```diff
@@ -470,6 +499,24 @@ protected int constructExpectedNumberOfTasks(DatastreamGroup dg, List<String> in
...
+    int partitionsPerTask = resolveConfigWithMetadata(datastreamPartitions.getDatastreamGroup(),
+        CFG_PARTITIONS_PER_TASK, _partitionsPerTask);
+    String taskPrefix = datastreamPartitions.getDatastreamGroup().getTaskPrefix();
+    ElasticTaskAssignmentInfo elasticTaskAssignmentInfo =
+        new ElasticTaskAssignmentInfo(actualPartitionsPerTask, actualPartitionsPerTask > partitionsPerTask);
+
+    if (!_elasticTaskAssignmentInfoHashMap.containsKey(taskPrefix)) {
+      registerElasticTaskAssignmentMetrics(taskPrefix);
+    }
+    _elasticTaskAssignmentInfoHashMap.put(taskPrefix, elasticTaskAssignmentInfo);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:517]: Do we not need to remove the taskPrefix from this Map when not in use.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java [hunk 5]
```diff
@@ -58,11 +62,20 @@ public void setup() throws IOException {
...
     _embeddedZookeeper.startup();
     _zkClient = new ZkClient(zkConnectionString);
+    DynamicMetricsManager.createInstance(new MetricRegistry(), "TestStickyPartitionAssignment");
   }
 
   @AfterMethod
-  public void teardown() {
+  public void teardown() throws Exception {
     _embeddedZookeeper.shutdown();
+    // A hack to force clean up DynamicMetricsManager
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java:71]: Can you please explain the reason behind the hack to understand the rationale for future use? I'm not able to make sense from this code.

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java [hunk 6]
```diff
@@ -415,6 +428,8 @@ public void testElasticTaskPartitionAssignmentCorrectlyAdjustsNumTasks() {
     // Partition assignment should go through
     assignment = strategy.assignPartitions(assignment, partitionsMetadata);
     validatePartitionAssignment(assignment, partitions, maxPartitionsPerTask, numTasksNeeded);
```
Comment (vmaheshw) [inline datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java:430]: Not sure if there is a direct way, but one test can be to have 1 instance, 1 task-A, 1 partition; Check the new metrics; then remove the task and add task with another prefix and see if the metric for the first prefix is no longer emitted and is emitted for the second prefix.

 since this is emitted only by the leader, but the metric is a gauge (which continuously emit), on zk expiry, the leader-> follower will continue to emit, while the new leader will also emit, resulting in incorrect value. So, if you introduce a deregisterMetrics path, you can validate that extra metrics are not emitted when not required i.e. no tasks for the datastream. (This is tricky and may need more thought, but since this is very leader specific, we either need to fix the gauge value somehow or deregister itself)

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 7]
```diff
@@ -592,4 +639,39 @@ private int getNumTasksForDatastreamFromZK(String stream) {
...
+            && _elasticTaskAssignmentInfoHashMap.get(taskPrefix).getNeedsAdjustment()) ? 1.0 : 0.0;
+    DYNAMIC_METRICS_MANAGER.registerGauge(CLASS_NAME, taskPrefix, PARTITIONS_PER_TASK_NEEDS_ADJUSTMENT,
+        needsAdjustmentSupplier);
+  }
+
+  private static class ElasticTaskAssignmentInfo {
+    private int _actualPartitionsPerTask;
+    private boolean _needsAdjustment;
+
+    ElasticTaskAssignmentInfo(int actualPartitionsPerTask, boolean needsAdjustment) {
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:664]: [FYI: Copying this comment I got from Ahmed once for inner classes]

Relevant: I remember looking at [this](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fstackoverflow.com%2Fquestions%2F6264657%2Fwhy-make-private-inner-class-member-public-in-java&data=02%7C01%7Cvmaheshwari%40linkedin.com%7C93a5aac105654d87be6908d7ed448b1c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637238750965657244&sdata=MZSudl8GkJj0yCq7ATaRGzwUUrysy8w4GOo%2BUsR%2Bftk%3D&reserved=0) SO question a long while ago, and liking this [answer](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fstackoverflow.com%2Fquestions%2F6264657%2Fwhy-make-private-inner-class-member-public-in-java&data=02%7C01%7Cvmaheshwari%40linkedin.com%7C93a5aac105654d87be6908d7ed448b1c%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637238750965667242&sdata=Nj3VR5a5gF0Rz7uM7k73o42bDVvI8Dt4HuWs1jShI4w%3D&reserved=0):

This method can be made public in order to indicate that it's semantically public, despite the fact that compiler doesn't enforce visibility rules in this particular case.

Imagine that during some refactoring you need to make this inner class top-level. If this method is private, how would you decide whether it should be made public, or some more restrictive modifier should be used? Declaring method as public tells reader the intentions of original author - this method shouldn't be considered an implementation detail.

It's worth noting the top-voted answer said using public was "nonsense" (i.e. ineffective/inconsequential) nonetheless.

**Review by vmaheshw — COMMENTED** [review]
Please evaluate the de-registration of the metrics in this class.

File: datastream-testcommon/src/main/java/com/linkedin/datastream/testutil/DatastreamTestUtils.java (hunk 2 above)
Comment (somandal) [reply datastream-testcommon/src/main/java/com/linkedin/datastream/testutil/DatastreamTestUtils.java:200]: Yup

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 3 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:646]: MetricsAware doesn't expose a deregister path. EventProducer, LiKafkaTransportProvider, etc fall into this category. Do you propose adding one? Connectors use a common metrics class which exposes de-registration code paths.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 4 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:517]: I need to expose this map in StickyMulticastStrategy (where I can find out that a datastream group is missing). Partition assignment is called per group, so there is no way to know that a datastream group is deleted. Is that okay with you?

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java (hunk 5 above)
Comment (somandal) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java:71]: I got this from another test, so don't know the rationale, but I do remember there was an issue with some tests getting older metrics, and I think this was added to remove that. I can double check what this is all about if you'd like.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 7 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:664]: So are you suggesting I make this class public?

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java (hunk 6 above)
Comment (somandal) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java:430]: We'd either need to extend MetricsAware or extend the AssignmentStrategy interface to add a method for deregistration of metrics. Is that okay with you? Also, depending on how this is done, will need to see how easy it is to simulate in tests.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 3 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:669]: As discussed offline, decided to go with ensuring the metric has the right value rather than deregistering it.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 4 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:540]: done

File: datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java (hunk 6 above)
Comment (somandal) [reply datastream-server/src/test/java/com/linkedin/datastream/server/assignment/TestStickyPartitionAssignment.java:430]: Don't need this test since we're not deregistering the metric, just resetting it. Which means the metric mbean will still exist, but have a value of 0.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 7 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java]: done

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 8]
```diff
@@ -154,6 +154,22 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
...
+  public Map<String, Set<DatastreamTask>> assign(List<DatastreamGroup> datastreams, List<String> instances,
+      Map<String, Set<DatastreamTask>> currentAssignment) {
+    // Remove entries for datastreams that aren't to be assigned in this round of task assignment. This resets the
+    // metrics correctly for datastreams that were deleted / stopped. This will be recalculated on further rounds for
+    // restarted datastreams.
+    Set<String> datastreamTaskPrefixes = datastreams.stream().map(DatastreamGroup::getTaskPrefix)
+        .collect(Collectors.toSet());
+    _elasticTaskAssignmentInfoHashMap.keySet().forEach(datastreamTaskPrefix -> {
+      if (!datastreamTaskPrefixes.contains(datastreamTaskPrefix)) {
+        _elasticTaskAssignmentInfoHashMap.remove(datastreamTaskPrefix);
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:167]: You can deregister the gauge as well.

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java [hunk 9]
```diff
@@ -450,6 +466,13 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
     return tasksToCleanUp;
   }
 
+  @Override
+  public void cleanupStrategy() {
+    // Clear the maintained state
+    _elasticTaskAssignmentInfoHashMap.clear();
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:472]: You can deregister the gauge metrics as well.

**Review by vmaheshw — DISMISSED** [review]

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 8 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:169]: done

File: datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java (hunk 9 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java:478]: donee

**Review by vmaheshw — APPROVED** [review]
LGTM

**Review by jzakaryan — APPROVED** [review]
